pkg/pmem-csi-driver/main.go

```diff
@@ -12,39 +12,45 @@ import (
-	clientKeyFile    = flag.String("clientKeyFile", "", "Client private key associated to client certificate, defaults to 'keyFile'")
+	flag.StringVar(&config.DriverName, "drivername", "pmem-csi.intel.com", "name of the driver")
+	flag.StringVar(&config.NodeID, "nodeid", "nodeid", "node id")
+	flag.StringVar(&config.Endpoint, "endpoint", "unix:///tmp/pmem-csi.sock", "PMEM CSI endpoint")
+	flag.BoolVar(&config.TestEndpoint, "testEndpoint", false, "also expose controller interface via endpoint (for testing only)")
+	flag.Var(&config.Mode, "mode", "driver run mode: controller or node")
+	flag.StringVar(&config.RegistryEndpoint, "registryEndpoint", "", "endpoint to connect/listen registry server")
+	flag.StringVar(&config.CAFile, "caFile", "", "Root CA certificate file to use for verifying connections")
+	flag.StringVar(&config.CertFile, "certFile", "", "SSL certificate file to use for authenticating client connections(RegistryServer/NodeControllerServer)")
+	flag.StringVar(&config.KeyFile, "keyFile", "", "Private key file associated to certificate")
+	flag.StringVar(&config.ClientCertFile, "clientCertFile", "", "Client SSL certificate file to use for authenticating peer connections, defaults to 'certFile'")
+	flag.StringVar(&config.ClientKeyFile, "clientKeyFile", "", "Client private key associated to client certificate, defaults to 'keyFile'")
 	/* Node mode options */
-	controllerEndpoint = flag.String("controllerEndpoint", "", "internal node controller endpoint")
-	deviceManager      = flag.String("deviceManager", "lvm", "device manager to use to manage pmem devices. supported types: 'lvm' or 'ndctl'")
-	showVersion        = flag.Bool("version", false, "Show release version and exit")
-	driverStatePath    = flag.String("statePath", "", "Directory path where to persist the state of the driver running on a node, defaults to /var/lib/<drivername>")
+	flag.StringVar(&config.ControllerEndpoint, "controllerEndpoint", "", "internal node controller endpoint")
+	flag.StringVar(&config.DeviceManager, "deviceManager", "lvm", "device manager to use to manage pmem devices. supported types: 'lvm' or 'ndctl'")
+	flag.StringVar(&config.StateBasePath, "statePath", "", "Directory path where to persist the state of the driver running on a node, defaults to /var/lib/<drivername>")
 
 	/* scheduler options */
-	schedulerListenAddr = flag.String("schedulerListen", "", "listen address (like :8000) for scheduler extender and mutating webhook, disabled by default")
+	flag.StringVar(&config.schedulerListen, "schedulerListen", "", "listen address (like :8000) for scheduler extender and mutating webhook, disabled by default")
 
-	version = "unknown" // Set version during build time
-)
+	/* metrics options */
+	flag.StringVar(&config.metricsListen, "metricsListen", "", "listen address (like :8001) for prometheus metrics endpoint, disabled by default")
```

Comment:
I'm undecided whether this really should be a separate port. It might be simpler to have a single HTTPS server for unknown clients which serves both scheduler and metrics endpoints.

Thoughts?

Reply:
I think I prefer to keep it separate. I asked in Kubernetes-CSI about the confidentiality of the metrics data and the statement there was that Google's own security team did not consider the data exposed by the sidecars confidential, i.e. exposing our own metrics data to anyone also should be fine. However, if we ever decided to require authentication of the client, then it is easier to do with a separate port because then we can do client checks already at the TLS level.